runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java

```diff
@@ -1721,20 +1729,20 @@ public Iterable<V> read() {
             return Collections.emptyList();
           }
 
-          if (localRemovals.contains(structuralKey)) {
+          if (keyState.removedLocally) {
             // this key has been removed locally but the removal hasn't been sent to windmill,
             // thus values in windmill(if any) are obsolete, and we only care about local values.
-            return Iterables.unmodifiableIterable(localAdditions.get(structuralKey));
+            return Iterables.unmodifiableIterable(keyState.localAdditions);
           }
           if (keyState.valuesCached || complete) {
             return Iterables.unmodifiableIterable(
-                Iterables.concat(keyState.values, localAdditions.get(structuralKey)));
+                Iterables.concat(keyState.values, keyState.localAdditions));
           }
           Future<Iterable<V>> persistedData = getFutureForKey(key);
           try (Closeable scope = scopedReadState()) {
             final Iterable<V> persistedValues = persistedData.get();
             if (Iterables.isEmpty(persistedValues)) {
-              Collection<V> local = localAdditions.get(structuralKey);
+              Collection<V> local = keyState.localAdditions;
               if (local.isEmpty()) {
                 // empty in both cache and windmill, remove key from cache.
                 keyStateMap.remove(structuralKey);
```

Comment:
Done. 

In the current impl, get on non exist keys leaves tombstones in keyStateMap, and those will be cleared in `persistDirectly`. But note that if the user never add/remove from the multimap, `persistDirectly` will not try to clear those tombstones, because we iterate through the keys only if `(hasLocalRemovals || hasLocalAdditions)`, so if the user never add/remove and keep `get()` on different nonexistent keys, the memory could grow unbound. I think this is a rare edge case that we can live with, wdyt?

Reply:
Yeah calling get on lots of keys but then not adding/removing seems unlikely
If it gets to large we'll just stop caching it too